cgen: C error when the function returns mut fixed array

### Describe the bug

```v
pub fn foo(mut fixed_array [3]u8) [3]u8 {
	return fixed_array
}

fn main() {
	mut fixed := [3]u8{}
	mut res := foo(mut fixed)
	println(res)
}
```

```
==================
              ^
/tmp/v_501/a.1882641659735005790.tmp.c:12674:22: error: initializing '_v_Array_fixed_u8_3' (aka 'struct _v_Array_fixed_u8_3') with an expression of incompatible type 'Array_fixed_u8_3' (aka 'unsigned char[3]')
        _v_Array_fixed_u8_3 _t1 = *fixed_array;
                            ^     ~~~~~~~~~~~~
1 warning and 1 error generated.
...
```

### Reproduction Steps

ditto

### Expected Behavior

no errors.

### Current Behavior

C error.

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

latest master

### Environment details (OS name and version, etc.)

all
> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -5284,7 +5284,7 @@ fn (mut g Gen) return_stmt(node ast.Return) {
 			g.autofree_scope_vars(node.pos.pos - 1, node.pos.line_nr, true)
 			g.write('return ')
 		}
-		if expr0.is_auto_deref_var() {
+		if expr0.is_auto_deref_var() && !fn_return_is_fixed_array {
 			if g.fn_decl.return_type.is_ptr() {
 				var_str := g.expr_string(expr0)
 				g.write(var_str.trim('&'))
@@ -5303,7 +5303,12 @@ fn (mut g Gen) return_stmt(node ast.Return) {
 				if fn_return_is_fixed_array && !node.types[0].has_flag(.option) {
 					g.writeln('{0};')
 					if node.exprs[0] is ast.Ident {
-						g.write('memcpy(${tmpvar}.ret_arr, ${g.expr_string(node.exprs[0])}, sizeof(${g.typ(node.types[0])})) /*ret*/')
+						typ := if expr0.is_auto_deref_var() {
+							node.types[0].deref()
+						} else {
+							node.types[0]
+						}
+						g.write('memcpy(${tmpvar}.ret_arr, ${g.expr_string(node.exprs[0])}, sizeof(${g.typ(typ)})) /*ret*/')
 					} else if node.exprs[0] in [ast.ArrayInit, ast.StructInit] {
 						if node.exprs[0] is ast.ArrayInit && node.exprs[0].is_fixed
 							&& node.exprs[0].has_init {

--- a/vlib/v/tests/return_fixed_array_test.v
+++ b/vlib/v/tests/return_fixed_array_test.v
@@ -21,3 +21,14 @@ fn test_without_alias() {
 	a := return_fixed_array()
 	assert a == [1, 2, 3]!
 }
+
+// for issue 20366: returns mut fixed array
+fn returns_mut_fixed_array(mut fixed_array [3]int) [3]int {
+	return fixed_array
+}
+
+fn test_returns_mut_fixed_array() {
+	mut fixed := [3]int{}
+	res := returns_mut_fixed_array(mut fixed)
+	assert res == [0, 0, 0]!
+}

